README.md
<h1 align="center">
  asdf-terragrunt
</h1>

<!-- Description section -->
<p align="center">
  <strong>a thin wrapper for Terraform that provides extra tools for working with multiple Terraform modules.</strong>
</p>

<!-- Badges section -->
<p align="center">
  <a href="https://github.com/kc-workspace/asdf-terragrunt/actions/workflows/main.yml">
    <img
      alt="github-main"
      src="https://img.shields.io/github/actions/workflow/status/kc-workspace/asdf-terragrunt/main.yml?style=flat-square&logo=github">
  </a>
  <a href="https://github.com/kc-workspace/asdf-terragrunt/releases">
    <img
      alt="version"
      src="https://img.shields.io/github/v/release/kc-workspace/asdf-terragrunt?style=flat-square&logo=github">
  </a>
  <a href="https://github.com/kc-workspace/asdf-terragrunt/commits/main">
    <img
      alt="version"
      src="https://img.shields.io/github/last-commit/kc-workspace/asdf-terragrunt/main?style=flat-square&logo=github">
  </a>
</p>

<!-- Links section -->
<h3 align="center">
  <a href="https://terragrunt.gruntwork.io/">terragrunt</a>
  <span> · </span>
  <a href="https://asdf-vm.com">asdf</a>
</h3>

> This is an asdf-vm plugin generated from [template][template-gh].

## Before start

> If you still see this section, mean this plugin is not ready yet

There are several things template cannot generate for you,
below are a list of thing we should do:

1. make sure that your GitHub repository already exist at [kc-workspace/asdf-terragrunt][plugin-gh]
2. please read [plugins create section][asdf-create-plugin] for more information
3. remove `before start` section once you completed

## Install

Plugin:

```sh
asdf plugin add "terragrunt" "https://github.com/kc-workspace/asdf-terragrunt.git"
```

App:

```sh
# Show all installable versions
asdf list all terragrunt

# Install latest version
asdf install terragrunt latest

# Set a version globally
asdf global terragrunt latest
# Set a version locally
asdf local terragrunt latest

# Now terragrunt commands are available
terragrunt
```

Check the [asdf][asdf-link] readme for instructions on
how to install & manage versions.

## Features

Plugins generated from asdf-plugin-template repository will
contains several extra features for every user including all below.

- `$DEBUG=<any-string>` to enabled debug mode (debug logs will guarantee to show regardless of other settings)
- `$ASDF_FORCE_DOWNLOAD=<any-string>` to always download even cache exist
- `$ASDF_INSECURE=<any-string>` to disable security features (e.g. checksum)
- `$ASDF_NO_CHECK=<any-string>` to disable pre-check features (e.g. check-cmd)
- `$ASDF_OVERRIDE_REF_REPO=<repository>` to override git repository URL when install with ref mode
- `$ASDF_OVERRIDE_OS=<os>` to override os name
- `$ASDF_OVERRIDE_ARCH=<arch>` to override arch name
- `$ASDF_OVERRIDE_EXT=<ext>` to override download extension
- `$GITHUB_TOKEN=<token>` to pass GitHub token on http request
- `$ASDF_LOG_FORMAT=<format>` to custom log format, there are several variables
  - **{datetime}** - for current datetime
  - **{date}** - for current date
  - **{time}** - for current time
  - **{level}** - for log level (always be 3 characters uppercase)
  - **{namespace}** - for formatted namespace (always have same length)
  - **{ns}** - for raw namespace (no formatting applied)
  - **{message}** - for log message
- `$ASDF_LOG_QUIET=<any-string>` - to disable only info logs
- `$ASDF_LOG_SILENT=<any-string>` - to disable all logs (including warning and error)

### Addition Features

The plugins might contains additional features
in addition to default features above.
You can take a look at [README.plugin.md][app-readme-md]

## Contributors

Read [CONTRIBUTING.md][contributing-md] file for more detail.

<!-- LINKS SECTION -->

[app-readme-md]: ./README.plugin.md
[contributing-md]: ./CONTRIBUTING.md
[plugin-gh]: https://github.com/kc-workspace/asdf-terragrunt
[template-gh]: https://github.com/kc-workspace/asdf-plugin-template
[asdf-link]: https://github.com/asdf-vm/asdf
[asdf-create-plugin]: https://asdf-vm.com/plugins/create.html


CHANGELOG.md
# Changelog

## [1.0.1](https://github.com/kc-workspace/asdf-terragrunt/compare/v1.0.0...v1.0.1) (2023-12-03)


### Improvements

* update plugin from template [autocommit] ([5aae571](https://github.com/kc-workspace/asdf-terragrunt/commit/5aae5714c611d3bafcfea2cba9bd544f529c616a))

## 1.0.0 (2023-11-06)


### Features

* **core:** applied template ([c2fb183](https://github.com/kc-workspace/asdf-terragrunt/commit/c2fb183026160cc14b9c4279f545ed34bf565ce3))
* **core:** applied template ([5012c90](https://github.com/kc-workspace/asdf-terragrunt/commit/5012c90caa048b35f49c5db97b878dbde727d545))
* **core:** applied template ([e0023d3](https://github.com/kc-workspace/asdf-terragrunt/commit/e0023d3d2172df4b3fba52bab8b9cee0983b4cac))
* **core:** applied template ([2c9bdfe](https://github.com/kc-workspace/asdf-terragrunt/commit/2c9bdfe2913149c298b843b0c7b7bc2576edf4cf))
* **core:** applied template ([b52946d](https://github.com/kc-workspace/asdf-terragrunt/commit/b52946d0891a835dbfea4d1bb134b1b755e8594f))
* initiate asdf plugin ([ef95ee8](https://github.com/kc-workspace/asdf-terragrunt/commit/ef95ee8a633b420efd8a7588b91a86b859ac9d59))


### Improvements

* update plugin from template [autocommit] ([2e45772](https://github.com/kc-workspace/asdf-terragrunt/commit/2e45772b141359099b4689043d189e4bed5aeb7a))
* update plugin from template [autocommit] ([44e70b5](https://github.com/kc-workspace/asdf-terragrunt/commit/44e70b5604a30f0de936be1c5066bc0a3c60b5e9))
* update plugin from template [autocommit] ([d5ae62f](https://github.com/kc-workspace/asdf-terragrunt/commit/d5ae62fb8dbbd2ac6fa57bf4d0d565d719993e6f))
* update plugin from template [autocommit] ([783878c](https://github.com/kc-workspace/asdf-terragrunt/commit/783878c3da7cc3b6818f401a06457e08e258f783))
* update plugin from template [autocommit] ([1efba22](https://github.com/kc-workspace/asdf-terragrunt/commit/1efba2208c8cd95352d4d579343575c3c5820ba0))
* update plugin from template [autocommit] ([297061e](https://github.com/kc-workspace/asdf-terragrunt/commit/297061eced46e31816219e5df4e7b6a8213b0e19))
* update plugin from template [autocommit] ([a33ff36](https://github.com/kc-workspace/asdf-terragrunt/commit/a33ff360583f5f962b92591ae40b48d8943b469a))


### Miscellaneous

* use asdf instead of ASDF ([52b2fb2](https://github.com/kc-workspace/asdf-terragrunt/commit/52b2fb214fc319f338dc90bd4413eb0051eac041))


lib/commands/command-test.bash
#!/usr/bin/env bash

## https://asdf-vm.com/plugins/create.html#extension-commands-for-asdf-cli

terragrunt --version


lib/common/main.sh
#!/usr/bin/env bash

## Main function for bin/scripts

## Get current OS name
## usage: `kc_asdf_get_os`
## variable:
##   - ASDF_OVERRIDE_OS for override arch
kc_asdf_get_os() {
  local ns="os.main"
  local os="${ASDF_OVERRIDE_OS:-}"
  if [ -n "$os" ]; then
    kc_asdf_warn "$ns" "user overriding OS to '%s'" "$os"
    printf "%s" "$os"
    return 0
  fi

  os="$(uname | tr '[:upper:]' '[:lower:]')"

  if command -v _kc_asdf_custom_os >/dev/null; then
    local tmp="$os"
    os="$(_kc_asdf_custom_os "$os")"
    kc_asdf_debug "$ns" "developer has custom OS name from %s to %s" "$tmp" "$os"
  fi

  printf "%s" "$os"
}

## Is current OS is macOS
## usage: `kc_asdf_is_darwin`
kc_asdf_is_darwin() {
  local os="${KC_ASDF_OS}" custom=""
  local darwin="${custom:-darwin}"
  [[ "$os" == "$darwin" ]]
}

## Is current OS is LinuxOS
## usage: `kc_asdf_is_linux`
kc_asdf_is_linux() {
  local os="${KC_ASDF_OS}" custom=""
  local linux="${custom:-linux}"
  [[ "$os" == "$linux" ]]
}

## Get current Arch name
## usage: `kc_asdf_get_arch`
## variable:
##   - ASDF_OVERRIDE_ARCH for override arch
kc_asdf_get_arch() {
  local ns="arch.main"
  local arch="${ASDF_OVERRIDE_ARCH:-}"
  if [ -n "$arch" ]; then
    kc_asdf_warn "$ns" "user overriding arch to '%s'" "$arch"
    printf "%s" "$arch"
    return 0
  fi

  arch="$(uname -m)"
  case "$arch" in
  aarch64*)
    arch="arm64"
    ;;
  armv5*)
    arch="arm"
    ;;
  armv6*)
    arch="arm"
    ;;
  armv7*)
    arch="arm"
    ;;
  i386)
    arch="386"
    ;;
  i686)
    arch="386"
    ;;
  x86)
    arch="386"
    ;;
  x86_64)
    arch="amd64"
    ;;
  esac

  if command -v _kc_asdf_custom_arch >/dev/null; then
    local tmp="$arch"
    arch="$(_kc_asdf_custom_arch "$arch")"
    kc_asdf_debug "$ns" "developer has custom ARCH name from %s to %s" "$tmp" "$arch"
  fi

  printf "%s" "$arch"
}

## Get download mode based on input filename
## usage: `kc_asdf_download_mode 'test.tar.gz'`
## output: file|archive|package
kc_asdf_download_mode() {
  local ns="download-mode.main"
  local filename="$1"
  local mode="file"

  

  

  kc_asdf_debug "$ns" "download mode of %s is %s" \
    "$filename" "$mode"
  printf "%s" "$mode"
}

## Check shasum of input path
## usage: `kc_asdf_checksum '/tmp/hello.tar.gz' 'https://example.com'`
## variables:
##   - ASDF_INSECURE for disable checksum verify
kc_asdf_checksum() {
  local ns="checksum.main"
  local filepath="$1" cs_url="$2"

  [ -n "${ASDF_INSECURE:-}" ] &&
    kc_asdf_warn "$ns" "Skipped checksum because user disable security" &&
    return 0

  local cs_tmp="checksum.tmp" cs_txt="checksum.txt"
  local dirpath filename
  dirpath="$(dirname "$filepath")"
  filename="$(basename "$filepath")"

  local cs_tmppath="$dirpath/$cs_tmp" cs_path="$dirpath/$cs_txt"

  kc_asdf_debug "$ns" "downloading checksum of %s at '%s'" \
    "$filename" "$cs_url"
  if ! kc_asdf_fetch_file "$cs_url" "$cs_tmppath"; then
    return 1
  fi

  kc_asdf_debug "$ns" "modifying checksum '%s' to '%s'" \
    "$cs_tmppath" "$cs_path"
  if command -v _kc_asdf_custom_checksum >/dev/null; then
    kc_asdf_debug "$ns" "use custom function to update checksum file"
    _kc_asdf_custom_checksum "$filename" "$cs_tmppath" "$cs_path"
  else
    if ! grep "$filename" "$cs_tmppath" >"$cs_path"; then
      kc_asdf_error "$ns" "missing %s on checksum file (%s)" \
        "$filename" "$cs_tmppath"
      return 1
    fi
  fi

  local cs_algorithm="256"
  local shasum="sha${cs_algorithm}sum"
  command -v "$shasum" >/dev/null ||
    shasum="shasum"

  local tmp="$PWD"
  cd "$dirpath" &&
    kc_asdf_exec "$shasum" --check "$cs_txt" >/dev/null &&
    cd "$tmp" || return 1
}

## Check GPG value from input path
## usage: `kc_asdf_gpg '/tmp/hello.tar.gz' 'https://example.com'`
kc_asdf_gpg() {
  local ns="gpg.main"
  # TODO: implement gpg verify
  kc_asdf_warn "$ns" "gpg verify is not implemented yet"

  ## Get GPG User ID from public key
  # gpg --list-packets aws.pub | grep -E '^:user ID packet: ' | sed 's|^:user ID packet: ||' | tr -d '"'
  return 0
}

## Get latest tags from GitHub
## usage: `kc_asdf_gh_latest`
kc_asdf_gh_latest() {
  local ns="gh-latest.main"
  local repo="$KC_ASDF_APP_REPO"
  local url="" version=""

  [[ "$repo" =~ ^https://github.com ]] ||
    return 1

  url="$(kc_asdf_fetch_location "$repo/releases/latest")"

  kc_asdf_debug "$ns" "fetch latest url: %s" "$url"
  if [ -n "$url" ] && [[ "$url" != "$repo/releases" ]]; then
    version="$(printf "%s\n" "$url" | sed 's|.*/tag/v\{0,1\}||')"
  fi

  kc_asdf_debug "$ns" "latest version is '%s'" "$version"
  [ -n "$version" ] &&
    printf "%s" "$version"
}

## List all tags from Git
## usage: `output_file="$(kc_asdf_tags_list)"`
kc_asdf_tags_list() {
  local ns="tags-list.main"
  local repo="$KC_ASDF_APP_REPO"
  local output
  output="$(kc_asdf_temp_file)"

  kc_asdf_debug "$ns" "querying from %s" "$repo"
  if git ls-remote --tags --refs "$repo" |
    grep -o 'refs/tags/.*' |
    cut -d/ -f3- >"$output"; then
    printf "%s" "$output"
    return 0
  fi

  kc_asdf_error "$ns" "listing failed (%s)" "$repo"
  return 1
}

## Filter only stable tags from tags list
## usage: `output_file="$(kc_asdf_tags_stable "$input_file")"`
kc_asdf_tags_stable() {
  local ns="tags-stable.main"
  local input="$1" output
  output="$(kc_asdf_temp_file)"
  local query='(-src|-dev|-latest|-stm|[-\.]rc|-alpha|-beta|[-\.]pre|-next|snapshot|master)'

  kc_asdf_debug "$ns" "filtering from %s" "$input"
  if [ -f "$input" ] && grep -ivE "$query" "$input" >"$output"; then
    __asdf_if_not_debug rm "$input"
    printf "%s" "$output"
    return 0
  fi

  kc_asdf_error "$ns" "filtering '%s' failed (%s)" "$query" "$input"
  return 1
}

## Sorting tags using semver
## usage: `output_file="$(kc_asdf_tags_sort "$input_file")"`
kc_asdf_tags_sort() {
  local ns="tags-sort.main"
  local input="$1" output
  output="$(kc_asdf_temp_file)"

  kc_asdf_debug "$ns" "sorting from %s" "$input"
  if [ -f "$input" ] &&
    sed 'h; s/[+-]/./g; s/.p\([[:digit:]]\)/.z\1/; s/$/.z/; G; s/\n/ /' "$input" |
    LC_ALL=C sort -t. -k 1,1 -k 2,2n -k 3,3n -k 4,4n -k 5,5n | awk '{print $2}' >"$output"; then
    __asdf_if_not_debug rm "$input"
    printf "%s" "$output"
    return 0
  fi

  kc_asdf_error "$ns" "sorting failed (%s)" "$input"
  return 1
}

## Filter only tag with input regex
## usage: `output_file="$(kc_asdf_tags_only "$input_file" ^v)"`
kc_asdf_tags_only() {
  local ns="tags-only.main"
  local input="$1" output
  output="$(kc_asdf_temp_file)"
  local regex="${2:-^\\s*v}"

  kc_asdf_debug "$ns" "filtering from %s" "$input"
  if [ -f "$input" ] &&
    grep -iE "$regex" "$input" >"$output"; then
    __asdf_if_not_debug rm "$input"
    printf "%s" "$output"
    return 0
  fi

  kc_asdf_error "$ns" "filtering '%s' failed (%s)" "$regex" "$input"
  return 1
}

## Formatting tags by remove input regex
## usage: `output_file="$(kc_asdf_tags_format "$input_file" ^v)"`
kc_asdf_tags_format() {
  local ns="tags-format.main"
  local input="$1" output
  output="$(kc_asdf_temp_file)"
  local regex="${2:-^\\s*v}"

  kc_asdf_debug "$ns" "formating from %s" "$input"
  if [ -f "$input" ] &&
    sed "s/$regex//" "$input" >"$output"; then
    __asdf_if_not_debug rm "$input"
    printf "%s" "$output"
    return 0
  fi

  kc_asdf_error "$ns" "formating '%s' failed (%s)" "$regex" "$input"
  return 1
}


lib/addon/system.sh
#!/usr/bin/env bash

## Get current OS name
## usage: `kc_asdf_get_os`
## variable:
##   - ASDF_OVERRIDE_OS for override arch
kc_asdf_get_os() {
  local ns="os.addon"
  local os="${ASDF_OVERRIDE_OS:-}"
  if [ -n "$os" ]; then
    kc_asdf_warn "$ns" "user overriding OS to '%s'" "$os"
    printf "%s" "$os"
    return 0
  fi

  os="$(uname | tr '[:upper:]' '[:lower:]')"

  if command -v _kc_asdf_custom_os >/dev/null; then
    local tmp="$os"
    os="$(_kc_asdf_custom_os "$os")"
    kc_asdf_debug "$ns" "developer has custom OS name from %s to %s" "$tmp" "$os"
  fi

  printf "%s" "$os"
}

## Is current OS is macOS
## usage: `kc_asdf_is_darwin`
kc_asdf_is_darwin() {
  local os="${KC_ASDF_OS}" custom=""
  local darwin="${custom:-darwin}"
  [[ "$os" == "$darwin" ]]
}

## Is current OS is LinuxOS
## usage: `kc_asdf_is_linux`
kc_asdf_is_linux() {
  local os="${KC_ASDF_OS}" custom=""
  local linux="${custom:-linux}"
  [[ "$os" == "$linux" ]]
}

## Get current Arch name
## usage: `kc_asdf_get_arch`
## variable:
##   - ASDF_OVERRIDE_ARCH for override arch
kc_asdf_get_arch() {
  local ns="arch.addon"
  local arch="${ASDF_OVERRIDE_ARCH:-}"
  if [ -n "$arch" ]; then
    kc_asdf_warn "$ns" "user overriding arch to '%s'" "$arch"
    printf "%s" "$arch"
    return 0
  fi

  arch="$(uname -m)"
  case "$arch" in
  aarch64*)
    arch="arm64"
    ;;
  armv5*)
    arch="arm"
    ;;
  armv6*)
    arch="arm"
    ;;
  armv7*)
    arch="arm"
    ;;
  i386)
    arch="386"
    ;;
  i686)
    arch="386"
    ;;
  x86)
    arch="386"
    ;;
  x86_64)
    arch="amd64"
    ;;
  esac

  if command -v _kc_asdf_custom_arch >/dev/null; then
    local tmp="$arch"
    arch="$(_kc_asdf_custom_arch "$arch")"
    kc_asdf_debug "$ns" "developer has custom ARCH name from %s to %s" "$tmp" "$arch"
  fi

  printf "%s" "$arch"
}

## System information
KC_ASDF_OS="$(kc_asdf_get_os)"
KC_ASDF_ARCH="$(kc_asdf_get_arch)"
export KC_ASDF_OS KC_ASDF_ARCH


lib/common/index.sh
#!/usr/bin/env bash

## Environment variables
## https://asdf-vm.com/plugins/create.html#environment-variables-overview

## General information
KC_ASDF_RES_PATH="${KC_ASDF_PLUGIN_PATH:?}/res"
KC_ASDF_ADDON_LIST=""
export KC_ASDF_RES_PATH KC_ASDF_ADDON_LIST

## Plugin information
KC_ASDF_ORG="kc-workspace"
KC_ASDF_NAME="asdf-terragrunt"
KC_ASDF_REPO="https://github.com/kc-workspace/asdf-terragrunt"
export KC_ASDF_ORG KC_ASDF_NAME KC_ASDF_REPO

## Application information
KC_ASDF_APP_TYPE="standalone"
KC_ASDF_APP_NAME="terragrunt"
KC_ASDF_APP_WEBS="https://terragrunt.gruntwork.io/"
KC_ASDF_APP_DESC="a thin wrapper for Terraform that provides extra tools for working with multiple Terraform modules."
KC_ASDF_APP_REPO="https://github.com/gruntwork-io/terragrunt"
export KC_ASDF_APP_TYPE KC_ASDF_APP_NAME
export KC_ASDF_APP_WEBS KC_ASDF_APP_DESC
export KC_ASDF_APP_REPO

# shellcheck source-path=SCRIPTDIR/internal.sh
source "${KC_ASDF_PLUGIN_PATH:?}/lib/common/internal.sh" || exit 1
# shellcheck source-path=SCRIPTDIR/defaults.sh
source "${KC_ASDF_PLUGIN_PATH:?}/lib/common/defaults.sh" || exit 1


lib/common/defaults.sh
#!/usr/bin/env bash

## Print error message
## usage: `kc_asdf_error 'something went %s' 'wrong'`
kc_asdf_error() {
  kc_asdf_log "ERR" "$@"
}

## Print info message
## usage: `kc_asdf_info 'this is a %s message' 'info'`
kc_asdf_info() {
  kc_asdf_log "INF" "$@"
}

## Print debug message
## usage: `kc_asdf_debug 'this is a %s message' 'debug'`
kc_asdf_debug() {
  kc_asdf_when_debug \
    kc_asdf_log "DBG" "$@"
}

## Print message
## usage: `kc_asdf_printf "test %s" "1"`
kc_asdf_printf() {
  kc_asdf_log '' "$@"
}

## Print start input step in info message
kc_asdf_step_start() {
  local step="$1" format="$2"
  shift 2

  kc_asdf_info "%9s | %-10s | $format" \
    "starting" "$step" "$@"
}

## Print finished input step in info message
kc_asdf_step_success() {
  local step="$1" format="$2"
  shift 2

  kc_asdf_info "%9s | %-10s | $format" \
    "finished" "$step" "$@"
}

## Print help header message
kcs_asdf_help_header() {
  kc_asdf_printf "# %s" "$1"
}

## Print log message with format
## usage: `kc_asdf_log '$KEY' 'this is a %s' 'message'`
## format: '[$KEY] $message'
kc_asdf_log() {
  local key="$1" format="$2"
  shift 2

  [ -n "$key" ] &&
    printf "[%-3s] " "$key" >&2
  # shellcheck disable=SC2059
  printf "$format\n" "$@" >&2
}

## Print error message and exit the program
## usage: `kc_asdf_throw 1 'something went %s' 'wrong'`
kc_asdf_throw() {
  local code="${1:-1}"
  shift

  kc_asdf_error "$@"
  exit "$code"
}

## Execute input command if debug mode enabled
## usage: `kc_asdf_when_debug echo "hello debugger"`
kc_asdf_when_debug() {
  [ -z "${DEBUG:-}" ] &&
    return 0
  "$@"
}

## Execute input command if debug mode disabled
## usage: `kc_asdf_when_not_debug echo "hello normal user"`
kc_asdf_when_not_debug() {
  [ -n "${DEBUG:-}" ] &&
    return 0
  "$@"
}

## Execute input command (if dry-run is disabled)
## usage: `kc_asdf_exec echo 'run'`
## variables:
##   - DRY_RUN | DRYRUN | DRY
kc_asdf_exec() {
  if [ -n "${DRY_RUN:-}" ] ||
    [ -n "${DRYRUN:-}" ] ||
    [ -n "${DRY:-}" ]; then
    kc_asdf_printf "[DRY] %s" "$*"
    return 0
  fi
  "$@"
}

## Fetch redirected location from url
## e.g. `kc_asdf_fetch_location https://google.com`
kc_asdf_fetch_location() {
  local max_redirs=10 tmp_file
  tmp_file="$(kc_asdf_temp_file)"
  export CURL_OPTIONS=(
    --head
    --write-out "%{url_effective}"
    --output "/dev/null"
  )
  export WGET_OPTIONS=(
    --server-response
    --spider
    --output-file "$tmp_file"
  )

  if ! kc_asdf_fetch "$1"; then
    unset CURL_OPTIONS WGET_OPTIONS
    return 11
  elif [ -f "$tmp_file" ]; then
    sed -n -e "s|^[ ]*Location: *||p" <"$tmp_file" |
      tail -n1 &&
      rm "$tmp_file"
  fi

  unset CURL_OPTIONS WGET_OPTIONS
}

## Download file from input url
## e.g. `kc_asdf_fetch_file https://google.com /tmp/index.html`
kc_asdf_fetch_file() {
  local max_redirs=10
  local url="$1" filename="$2"
  export CURL_OPTIONS=(
    --output "$filename"
  )
  export WGET_OPTIONS=(
    --output-file "$filename"
  )

  if ! kc_asdf_fetch "$url"; then
    unset CURL_OPTIONS WGET_OPTIONS
    return 12
  fi
  unset CURL_OPTIONS WGET_OPTIONS
}

## Fetch data from url
## usage: `kc_asdf_fetch https://google.com`
## variables:
##   - CURL_OPTIONS=() for curl options
##   - WGET_OPTIONS=() for wget options
##   - GITHUB_API_TOKEN for authentication
kc_asdf_fetch() {
  local url="$1"
  local cmd="" options=()

  local token=""
  if [[ "$url" =~ ^https://github.com ]]; then
    token="${GITHUB_API_TOKEN:-}"
    [ -z "$token" ] && token="${GITHUB_TOKEN:-}"
    [ -z "$token" ] && token="${GH_TOKEN:-}"
  fi

  local max_redirs=10

  if command -v "curl" >/dev/null; then
    cmd="curl"
    options+=(
      --fail
      --silent
      --show-error
      --location
      --max-redirs "$max_redirs"
    )
    [ -n "${CURL_OPTIONS:-}" ] &&
      options+=("${CURL_OPTIONS[@]}")
    [ -n "$token" ] &&
      options+=(--header "Authorization: token $token")
  elif command -v "wget" >/dev/null; then
    cmd="wget"
    options+=(
      --quiet
      --max-redirect "$max_redirs"
    )
    [ -n "${WGET_OPTIONS:-}" ] &&
      options+=("${WGET_OPTIONS[@]}")
    [ -n "$token" ] &&
      options+=(--header "Authorization: token $token")
  fi

  if [ -n "$cmd" ]; then
    kc_asdf_debug "exec: %s %s %s" \
      "$cmd" "${options[*]}" "$url"
    if ! "$cmd" "${options[@]}" "$url"; then
      kc_asdf_error "fetching %s failed" "$url"
      return 10
    fi
  else
    kc_asdf_error "fetching command not found (e.g. curl, wget)"
    return 10
  fi
}

## Parse template with input values
## usage: `kc_asdf_template "{template}" 'template=hello'`
kc_asdf_template() {
  local kv key value
  local template="$1"
  shift
  for kv in "$@"; do
    key="${kv%%=*}"
    value="${kv##*=}"
    template="${template//\{$key\}/$value}"
  done
  printf "%s" "$template"
}

## Quick helper to throw error if install type is not support
## usage: `kc_asdf_install_not_support "ref"`
## variables:
##   - ASDF_INSTALL_TYPE - set by asdf core
kc_asdf_install_not_support() {
  for t in "$@"; do
    if [[ "$t" == "${ASDF_INSTALL_TYPE:?}" ]]; then
      kc_asdf_throw 7 "your install type (%s) is not supported by plugins" \
        "$t"
    fi
  done
}

## Get current OS name
## usage: `kc_asdf_get_os`
## variable:
##   - ASDF_OVERRIDE_OS for override arch
kc_asdf_get_os() {
  local os="${ASDF_OVERRIDE_OS:-}"
  if [ -n "$os" ]; then
    kc_asdf_info "user overriding OS to '%s'" "$os"
    printf "%s" "$os"
    return 0
  fi

  os="$(uname | tr '[:upper:]' '[:lower:]')"

  printf "%s" "$os"
}

## Get current Arch name
## usage: `kc_asdf_get_arch`
## variable:
##   - ASDF_OVERRIDE_ARCH for override arch
kc_asdf_get_arch() {
  local arch="${ASDF_OVERRIDE_ARCH:-}"
  if [ -n "$arch" ]; then
    kc_asdf_info "user overriding arch to '%s'" "$arch"
    printf "%s" "$arch"
    return 0
  fi

  arch="$(uname -m)"
  case "$arch" in
  aarch64*)
    arch="arm64"
    ;;
  armv5*)
    arch="arm"
    ;;
  armv6*)
    arch="arm"
    ;;
  armv7*)
    arch="arm"
    ;;
  i386)
    arch="386"
    ;;
  i686)
    arch="386"
    ;;
  x86)
    arch="386"
    ;;
  x86_64)
    arch="amd64"
    ;;
  esac

  printf "%s" "$arch"
}

## Extract compress file
## usage: `kc_asdf_extract /tmp/file.tar.gz /tmp/file`
kc_asdf_extract() {
  local input="$1" output="$2"
  local options=(
    -xzf
    "$input"
    -C "$output"
    --strip-components ""
  )

  kc_asdf_debug "exec: tar %s" \
    "${options[*]}"
  tar "${options[@]}"
}

## Check shasum of input txt
## usage: `kc_asdf_checksum /tmp/test/checksum.txt`
kc_asdf_checksum() {
  local input="$1"
  local dir file
  dir="$(dirname "$input")"
  file="$(basename "$input")"

  local bit="256"
  local cmd="sha${bit}sum"
  command -v "$cmd" >/dev/null ||
    cmd="shasum"

  kc_asdf_debug "exec: %s --check %s (%s)" \
    "$cmd" "$file" "$dir"

  local tmp="$PWD"
  cd "$dir" ||
    return 1
  "$cmd" --check "$file" >/dev/null ||
    return 1
  cd "$tmp" ||
    return 1
}

## Get latest tags from GitHub
## usage: `kc_asdf_gh_latest`
kc_asdf_gh_latest() {
  local repo="$KC_ASDF_APP_REPO"
  local url="" version=""

  [[ "$repo" =~ ^https://github.com ]] ||
    return 30

  url="$(kc_asdf_fetch_location "$repo/releases/latest")"

  kc_asdf_debug "github latest url: %s" "$url"
  if [ -n "$url" ] && [[ "$url" != "$repo/releases" ]]; then
    version="$(printf "%s\n" "$url" | sed 's|.*/tag/v\{0,1\}||')"
    kc_asdf_debug "use '%s' mode to resolve latest" "github"
  fi

  kc_asdf_debug "latest version is '%s'" "$version"
  if [ -n "$version" ]; then
    printf "%s" "$version"
  else
    return 31
  fi
}

## List all tags from Git
## usage: `output_file="$(kc_asdf_tags_list)"`
kc_asdf_tags_list() {
  local repo="$KC_ASDF_APP_REPO"
  local output
  output="$(kc_asdf_temp_file)"

  kc_asdf_debug "querying all tags from %s" \
    "$repo"
  if git ls-remote --tags --refs "$repo" |
    grep -o 'refs/tags/.*' |
    cut -d/ -f3- >"$output"; then
    printf "%s" "$output"
    return 0
  fi
  kc_asdf_error "cannot list tags from repo (%s)" \
    "$repo"
  return 20
}

## Filter only stable tags from tags list
## usage: `output_file="$(kc_asdf_tags_stable "$input_file")"`
kc_asdf_tags_stable() {
  local input="$1" output
  output="$(kc_asdf_temp_file)"
  local query='(-src|-dev|-latest|-stm|[-\.]rc|-alpha|-beta|[-\.]pre|-next|snapshot|master)'

  kc_asdf_debug "filtering stable tags from %s" \
    "$input"
  if [ -f "$input" ] &&
    grep -ivE "$query" "$input" >"$output"; then
    kc_asdf_when_not_debug rm "$input"
    printf "%s" "$output"
    return 0
  fi
  kc_asdf_error "filter stable tags failed (input=%s)" \
    "$input"
  return 24
}

## Sorting tags using semver
## usage: `output_file="$(kc_asdf_tags_sort "$input_file")"`
kc_asdf_tags_sort() {
  local input="$1" output
  output="$(kc_asdf_temp_file)"

  kc_asdf_debug "sorting tags from %s" \
    "$input"
  if [ -f "$input" ] &&
    sed 'h; s/[+-]/./g; s/.p\([[:digit:]]\)/.z\1/; s/$/.z/; G; s/\n/ /' "$input" |
    LC_ALL=C sort -t. -k 1,1 -k 2,2n -k 3,3n -k 4,4n -k 5,5n | awk '{print $2}' >"$output"; then
    kc_asdf_when_not_debug rm "$input"
    printf "%s" "$output"
    return 0
  fi
  kc_asdf_error "sort tags failed (input=%s)" \
    "$input"
  return 24
}

## Filter only tag with input regex
## usage: `output_file="$(kc_asdf_tags_only "$input_file" ^v)"`
kc_asdf_tags_only() {
  local input="$1" output
  output="$(kc_asdf_temp_file)"
  local regex="${2:-^\\s*v}"

  kc_asdf_debug "filtering tags with regex '%s' from %s" \
    "$regex" "$input"
  if [ -f "$input" ] &&
    grep -iE "$regex" "$input" >"$output"; then
    kc_asdf_when_not_debug rm "$input"
    printf "%s" "$output"
    return 0
  fi
  kc_asdf_error "filter tags failed (input=%s)" \
    "$input"
  return 24
}

## Formatting tags by remove input regex
## usage: `output_file="$(kc_asdf_tags_format "$input_file" ^v)"`
kc_asdf_tags_format() {
  local input="$1" output
  output="$(kc_asdf_temp_file)"
  local regex="${2:-^\\s*v}"

  kc_asdf_debug "formatting tags with regex '%s' from %s" \
    "$regex" "$input"
  if [ -f "$input" ] &&
    sed "s/$regex//" "$input" >"$output"; then
    kc_asdf_when_not_debug rm "$input"
    printf "%s" "$output"
    return 0
  fi
  kc_asdf_error "format tags failed (input=%s)" \
    "$input"
  return 24
}

## Create temp file and return path
## usage: `kc_asdf_temp_file`
kc_asdf_temp_file() {
  mktemp
}

## Create temp file and return path
## usage: `kc_asdf_temp_dir`
kc_asdf_temp_dir() {
  mktemp -d
}
